### ISSUE TITLE
Mixed Precision: which 'endpoint layer' to set dtype='float32' in Dueling DQN

### ISSUE THREAD

**Issue by turmeric-blend** [tensorflow/tensorflow#38703]
Hi, following the [mixed precision documentation](https://www.tensorflow.org/guide/keras/mixed_precision#overview) , it is recommended to have dtype of output layer as float32

>  And regardless of what your model ends in, make sure the output is float32.

However, I am not sure which 'layer' to apply float32 for Dueling DQN model. The architecture is as follows:
```
body = tf.keras.Sequential([tf.keras.layers.Dense(256, activation='relu'),
                             tf.keras.layers.Dense(256, activation='relu'),
                             tf.keras.layers.Dense(256, activation='relu'),])

value = tf.keras.Sequential([tf.keras.layers.Dense(256, activation='relu'),
                              tf.keras.layers.Dense(1),])

advantage = tf.keras.Sequential([tf.keras.layers.Dense(256, activation='relu'),
                                  tf.keras.layers.Dense(3, use_bias=False),])

q = body(training_input)

v = value(q)

a = advantage(q)
a -= tf.reduce_mean(a, axis=-1, keepdims=True)

final_q = v + a 
# final_q is used to calculate loss
```
should I set dtype='float32' at cases (1, 2, or 3):

1. value **&** advantage `Sequential` function
2. value `Sequential` function **&**`tf.reduce_mean()` function
3. **cast** final_q as float32

or maybe something else?

**Comment by reedwm** [comment]
Case 3: cast final_q to float32. Because the model does not end in softmax, the only important aspect is that the loss is done in float32, so final_q should be cast to float32.

FYI, there are two motivations in doing the loss in float32
1. Often, `tf.reduce_sum` is used to reduce a batch of losses into a single loss. If the batch size is very large, this can overflow in float16. With `Model.fit`, the reduce_sum is done in float32 by Keras, so this recommendation only applies to custom training loops.
2. If custom loss are used, the loss may not support float16.
